### PR TITLE
ROX-10836 add NetworkpoliciesStore to namespaceDispatcher listeners

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

The `NetworkPoliciesStore` needs to be added to the list of listener in the `namespaceDispatcher`.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

* [x] Unit tests passing
* [x] Added metrics to test manually.
    * Create multiple namespaces
    * Create multiple network policies in each namespace
    * Delete one of the namespaces and check the network policies in that namespace get deleted

### CODE REVIEW

File: sensor/kubernetes/listener/resources/networkpolicy_store.go
Comment (vikin91) [inline sensor/kubernetes/listener/resources/networkpolicy_store.go:206]:
```suggestion
	netpols, found := n.data[namespace]
	if !found {
		return
	}
```

**Review by vikin91 — APPROVED** [review]
Looks good (but please finish the tests you have under wip).

Just one suggestion regarding idiomatic check for checking presence of an element in the map.

File: sensor/kubernetes/listener/resources/networkpolicy_store.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_store.go:207]: Thank you! I completely missed that.

File: sensor/kubernetes/listener/resources/networkpolicy_store.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy_store.go:212]: nit: Shouldn't just this be enough to delete the whole thing? I assume that by using map's `delete` method you're anyway just dereferencing and the memory will be freed by garbage collection at some point. So maybe deleting the top-level reference would do the trick?

**Review by fredrb — APPROVED** [review]
Did this come up due to a bug or did you just realize that there were these `NamespaceDeletionListener` structs being passed to the namespace dispatcher and ours was missing?

I'd be curious to test if k8s would send deletion events to all child resources if the namespace gets deleted. 🤔

**Comment by lvalerom** [comment]
> Did this come up due to a bug or did you just realize that there were these `NamespaceDeletionListener` structs being passed to the namespace dispatcher and ours was missing?

We just realized that there were `NamespaceDeletionListener` in the namespace dispatcher.

> I'd be curious to test if k8s would send deletion events to all child resources if the namespace gets deleted. 🤔

@vikin91 @fredrb I've tested this manually creating multiple namespaces and multiple network policies in each namespace and from what I've seen k8s sends deletion events at least for child network policies. The thing is; I'm not sure if we can rely completely on k8s sending these events. Seeing that we have this `OnNamespaceDeleted` function for deployments, pods, and services I think it's best to be on the safe side and have it for network policies too until we figure this out.

PD: There is an e2e test failing but I don't think it's related to this PR since it's also failing in master.

File: sensor/kubernetes/listener/resources/networkpolicy_store.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_store.go:212]: To be honest I'm not sure how the garbage collector works in go. If we are sure it will delete the inner objects, I'm fine with deleting just the top-level reference.

File: sensor/kubernetes/listener/resources/networkpolicy_store.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_store.go:202]: I realized that when we clear a namespace completely the metrics will show one remaining object in the store. This is because we delete the top level object before sending the metrics update, therefore prometheus only gets updates in the remaining namespaces. It's a minor issue but something to take into consideration.

File: sensor/kubernetes/listener/resources/networkpolicy_store.go
Comment (vikin91) [reply sensor/kubernetes/listener/resources/networkpolicy_store.go:212]: I think keeping only `delete(n.data, namespace)` should work too, but deletion of namespaces is rare, so even if there is a penalty for iterating, it would not have any influence. I am fine, keeping it as it is.

**Comment by vikin91** [comment]
> The thing is; I'm not sure if we can rely completely on k8s sending these events

I am fine keeping it as it is now. As long as we `mutex.Lock`, we should be on the safe side.

**Comment by vikin91** [comment]
Can we merge this? I guess release guys may be waiting to close the milestone

**Comment by vikin91** [comment]
@JoukoVirtanen we are still waiting for the CI to finish, but this should be good to be merged. Feel free to merge this if this could block you in the release.

**Comment by lvalerom** [comment]
> Can we merge this? I guess release guys may be waiting to close the milestone
>
> @JoukoVirtanen we are still waiting for the CI to finish, but this should be good to be merged. Feel free to merge this if this could block you in the release.

Just waiting on the CI to be green. As soon as it finishes we can merge this.

**Pull request merged**
